### ISSUE TITLE
[CSV] Registration form sample file

### ISSUE THREAD

**Issue by mahalakshme** [avniproject/avni-webapp#1394]
### Need:
- Many new implementations that are getting onboarded need this. Currently only implementation team members with trial and effort can upload it. This will cost our partners more. So the idea here is to enable our partners itself to be able to upload their old data.

#### Person subject type:
- For registration create(Edit of an entry will not anymore be possible via CSV. We will support only create-not scope of this card to restrict):
      - Instead of 'Id column' add 'Id from previous system' as column with mentioned descriptor in the sample file
      -  Mention as 'Optional' for the fields that are not marked mandatory in the form.
      - If a registration form is associated with only one subject type, then subject type column should not get displayed. 
      - Move 'Date Of Registration' and 'Registration Location' to before 'First Name' like available in 'Data Entry App'. 
      - If 'Lowest Address levels' defined in the subject type, then the hierarchy until lowest 'lowest address level' is sufficient to be mentioned.
      - If multiple hierarchies are present, then only the hierarchies to which the defined 'Lowest Address level' belongs to need to be mentioned in the sample file.
      - For coded concepts, mention 'Allowed values: {}'
      - For coded concepts single select, mention 'Only single value allowed.'
      - For coded concepts multiple select, mention 'Format: Separate multiple values by a comma.'
      - For date concepts(including registration date) specify the format as 'Format: DD-MM-YYYY'
      -  For numeric concepts, mention 'Max value allowed:' or 'Min value allowed' if high or low absolute defined. Else mention 'Allowed values: Any number'
      -  For text and notes concepts, mention 'Any Text'      
      - Order of descriptions:
              1. Optional
              2. Allowed values/Max value, Min value/Format
              3. Only single value allowed/Separate multiple values by a comma/Default value:

Sample file for prerelease, APF Odisha, create mode, individual registration form, Person subject type: 
https://docs.google.com/spreadsheets/d/123ZLzB5xZnOSL95YetNVURHn-WZKYpXd1ctti8kr3lk/edit?usp=sharing

#### For Subject type of types, Individual, Group, Household 
- columns, Date Of Birth, Date Of Birth Verified, Gender, First Name, Last Name, should not get exported and instead Name column should get displayed. For Household, 'Total members' column should also get displayed

- Make sure to add necessary tests as needed
- Do in a manner such that for enrolment and encounter forms code is reusable.

### Technical:
Refer [this](https://github.com/avniproject/avni-server/pull/823/commits/8ecf552dd715e8c0e1cd122b44873209752af451) pull request for doing this.

### Out of scope:
- Upload of data
- The idea is to make the process easy for the most commonly used data types and iteratively improve for others as well.
- Edit of forms.
- For registration edit:
      - Add Legacy id and UUID as columns instead of Id column. (to prevent incorrect update of existing records)
      - 'Mandatory' field descriptor can be removed
       - For registration form, 'subject type' column should never get displayed. 
- Sample file for prerelease, APF Odisha, edit mode, individual registration form, Person subject type: 
https://docs.google.com/spreadsheets/d/15qgMLhA72-ta2ouC2uDDy_OvgDTj6U-tsaUUiys-cPY/edit?usp=sharing

### Questions:
Mandatory(might  be subject to value in other fields)

### Inputs:
- type of subject -done
- QG/RQG - other concepts, media concepts - other concepts-out of scope
- pull request - shraddha
- read only - fields might want to upload
- mandatory-done
- API - validation-done
- Mostly rule validations switched off
- once we 've idea
- performance testing
- start with integration tests
- rules exection - out of scope
- dry run
- edit?

**Comment by petmongrels** [comment]
- Address level type is not a large hierarchy for which we require custom sql. We should be able to do this in memory using loading all domain objects (i.e. ALT)
- Same for subject type
- lets rename SubjectHeaders (and others) to SubjectHeaderCreator. SubjectHeaders sounds like POJO it is actually a component.
- AbstractHeaders should take its dependency via constructor not auto-wiring directly as a field in the base class. Otherwise this object cannot be constructed by a unit test and the signature doesn't tell what the object needs.
- org.avni.server.service.ImportHelperService#assertNotNull when would this actually throw any error? There is no user input here. The header creation is a data consistent process. If not then lets remove this method and all the calls to it.
- org.avni.server.service.SubjectImportService#generateSampleFile (and other such methods) should not catch exception and throw runtimeexception. there is no value added here as no Checked exception is being thrown.

**Comment by mahalakshme** [comment]
@ombhardwajj Just realised on a call we have separate multiple lines by bar like below:

<img width="845" alt="Image" src="https://github.com/user-attachments/assets/4b69d6c9-9f88-4160-b5ba-ae13035f0637" />

It should be separated by just . and space. Eg: Optional. Only single value allowed. Allowed values: {1,2,3}

I should ve mentioned explicitly as AC. But since this was followed for sometime in Location and User and Catchment sample file as well, didnt realise that. Let me know if you prefer fixing this in a separate card.

**Comment by ombhardwajj** [comment]
Yes @mahalakshme it would be better if we keep it separate since we are currently blocked by [avni-server#833](https://github.com/avniproject/avni-server/issues/833)

**Comment by AchalaBelokar** [comment]
- [ ] Not able to download sample file of any orgnisation

**Comment by ombhardwajj** [comment]
Try again @AchalaBelokar  this seems to have been fixed by Vivek's commit yesterday

**Comment by AchalaBelokar** [comment]
program has to be provided always, like done in Subject type for Subject upload

**Comment by mahalakshme** [comment]
@AchalaBelokar for registration why is program needed?

**Comment by ombhardwajj** [comment]
@mahalakshme She meant that subject type needs to be provided always

**Comment by AchalaBelokar** [comment]
- [ ] Login with maha@iphcd and try to upload the csv without location as we mention in the csv file it is not mandatory it is throwing the error.
-

[sample-Subject---Family.csv](https://github.com/user-attachments/files/19754501/sample-Subject---Family.csv)

[errors-sample-Subject---Family.csv](https://github.com/user-attachments/files/19754514/errors-sample-Subject---Family.csv)

**Comment by ombhardwajj** [comment]
@AchalaBelokar You should add this comment in avniproject/avniserver#833

**Comment by AchalaBelokar** [comment]
- [ ] Mandatory field is showing as optional. I am login as maha@iphscd and download the sample there mandatory fields are optional

**Comment by himeshr** [comment]
> * [ ]  Mandatory field is showing as optional. I am login as maha@iphscd and download the sample there mandatory fields are optional

@AchalaBelokar and @mahalakshme have marked addressColumns as Mandatory always. The only issue i see is when we have a single Subject for which data can be captured across multiple addressLevel hierarchies. In which case, user would have to specify address for one of the hierarchies mandatorily.

**Comment by AchalaBelokar** [comment]
- [ ] Login with maha@gpart check the activity type csv checked mandatory and non mandatory fields.
- [ ] The inventory item csv showing mandatory field is showing as optional

[sample-Subject---Inventory Item.csv](https://github.com/user-attachments/files/19955212/sample-Subject---Inventory.Item.csv)

**Comment by AchalaBelokar** [comment]
- [ ] Login with beulah@jscsuat_new on prerelease then I download sample file of individual but after download the file in the form mandatory field is showing the last name but in the form it is showing us cast

[Individual upload.csv](https://github.com/user-attachments/files/19976869/Individual.upload.csv)

<img width="1440" alt="Image" src="https://github.com/user-attachments/assets/877754e8-2784-433a-8a70-8afea4820a89" />

**Comment by petmongrels** [comment]
bug description is not clear.

**Comment by himeshr** [comment]
> * [ ]  The issue is in the CSV file we have the field as first name and last name but in the registration form we have the full name and caste and if we upload the data with first name and last name . The data is created as first name and caste name
>   eg : in the csv if we gave [first name : Test] [last name : din0201] but in the DEA it display as [First name : Test] [Caste : din0201]



This is due to the fact that, in jscs and its uat / dev orgs, like jscsuat_new, we have a translation for  "lastName" => "Caste", in english language. This might be for business reasons to capture caste as part of lastName field, and use it accordingly in reports and everywhere else.

Ignore this issue and validate further.

**Comment by dinesh2096** [comment]
1.Can we have error saying subject type(if we gave family instead of participation) is incorrect instead saying the whole error Mandatory columns are missing in header from uploaded file - total members. Please refer to sample file for the list of mandatory headers. Unknown headers - current or temporary address, phone no, cohort id, gender, parents/spouse/guardian name, st category, vgkk hospital mrd no., please specify, other type of enrolment, last name, date of birth verified, type of enrolment, please specify other st category, marital status, id, social categories, test mutliple, please specify other adivasi tribe, adivasi tribe, max value, date of birth, please specify other social category included in file. Please refer to sample file for valid list of headers. we have say subject type is mismatched.
2. If the location is marked as optional and if we removed the lowest address level then we are getting the error saying the missing header (org : iphscd) if we use the already downloaded file 
3.If we download the new file then location shows both in mandatory and non mandatory.
4.If the select the group subject and upload the person subject type file it is creating the data instead of showing error (happens in vice versa)
5.Total members is acceptinng the negative values but we have limit for 1 in the ccv upload

**Comment by petmongrels** [comment]
1. This surely can be done, but we are currently trying to make the basic thing work. Once we have higher usage of this, we could consider offering intelligence like this.
2. We are not supporting optional address level right now
3. not a bug. the PHC is coming due it being first form element in the form :-)
4. fixed
5. Not part of subject upload card.

**Comment by dinesh2096** [comment]
Test case : 

- [ ] Tested the without giving data to mandatory field 
- [ ] Tested with giving data to mandatory field 
- [ ] Tested without giving data to particular mandatory field like first name and validate the error 
- [ ] Tested by removing the mandatory field and validate the error (subject type, Registration date ,frst name and Date og birth)
- [ ] Tested all the field are optional and file got upload without any error and validate cilent app and webapp is breaking 
- [ ] Tested the field upload csv file in other subject type and validate the error 
- [ ] Tested the location with lowest address level 
- [ ] Tested with hierarchy location 
- [ ] Verify that coded concept fields display 'Allowed values: {value list}' in the sample file.
- [ ] Verified that single select coded fields display 'Only single value allowed.
- [ ] Verified that multi-select coded fields display 'Format: Separate multiple values by a comma.
- [ ] Verified that date fields show 'Format: DD-MM-YYYY' in the sample file.
- [ ] Verified that numeric fields display constraints such as 'Max value allowed:', 'Min value allowed:', or 'Allowed values: Any number' based on configuration.
- [ ] Verified that text and notes fields display 'Any Text' in the descriptor.
- [ ] Verified that descriptors appear in this order: 'Optional' → 'Allowed values/Format' → 'Only single/Multiple values/Default value'.
- [ ] Verified that for Household subject type, 'Total members' column is included in the sample file.
- [ ] Verified that the sample file includes 'Id from previous system' column instead of 'Id' with appropriate descriptor.

QA Reference file : 

[sample-Subject---Participant (3).csv](https://github.com/user-attachments/files/20098795/sample-Subject---Participant.3.csv)

**Comment by dinesh2096** [comment]
Found a issue when we added the date birth in the CSV file but it is not reflected in the web app and cilent app 

execution id : 
30282
